src/benchmark.rs

```diff
@@ -29,7 +28,7 @@ fn thread_func(benchmark: Arc<Vec<Box<(dyn Runnable + Sync + Send)>>>, config: A
     context.insert("base".to_string(), Yaml::String(config.base.to_string()));
 
     for item in benchmark.iter() {
-      item.execute(&mut context, &mut responses, &mut reports, &config);
+      item.execute(&mut context, &mut responses, &mut reports, &config).await;
```

Comment:
I think the problem is that even though `item.execute()` is async, here we simply `await` on it effectively make this iteration a sequential process, only one future will be running at a time and the `await` yield control but only to find it has to yield back  immediately which incurred lots of useless context switches and slows things down a lot.

Reply:
I think we need to refactor drill to `await` on multiple futures instead of only one at a time.

Reply:
I want to bring to this conversation a bit of background why I started working on this https://github.com/fcsonline/drill/pull/33 to fix the performance issues described here https://github.com/fcsonline/drill/issues/25. Also I decided to include `async` in the same branch.

The main point here is that some people expect drill to work like `jmeter`, with threads and interactions (one by one) and other people expect to work at maximum throughput like `ab`. A benchmark with steps can have dependencies between them and sometimes the `ab` is not feasible.

The question you exposing in the previous comment remembers me the point where I defined the concept of modes:

https://github.com/fcsonline/drill/blob/e765fc13fcf1154d2972c2af5938cf5b799d3e93/README.md#modes

to differentiate how do you want drill to work. I'm not saying we need to go in this direction. It was a change so massive that it reached a point hard to merge it.

I prefer to reach a point where we have `async` branch is ready to be merge rather than a huge refactor as I started and evaluate if we want the concept of modes in future versions. The only requirement here is that the performance in `async` branch and in `master` should be comparable leaving aside the way how iterations are executed.